tt_metal/impl/dispatch/kernels/remote_completion_queue_writer.cpp

```diff
@@ -0,0 +1,90 @@
+        volatile tt_l1_ptr uint32_t *buffer_transfer_command_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(buffer_transfer_start_addr);
+        uint32_t is_program = header->is_program_buffer;
+        uint32_t num_pages = header->num_pages;
+        const uint32_t dst_buf_type = buffer_transfer_command_ptr[5];
+        bool reading_buffer = (!is_program) & (num_pages > 0 & (BufferType)dst_buf_type == BufferType::SYSTEM_MEMORY);
+
+        uint32_t completion_data_size = header->completion_data_size;
+        completion_queue_reserve_back(completion_data_size);
+        write_event(uint32_t(&header->event));
+
+        uint32_t wrap = header->wrap;
+        if ((DeviceCommand::WrapRegion)wrap == DeviceCommand::WrapRegion::COMPLETION) {
+            cq_write_interface.completion_fifo_wr_ptr = completion_queue_start_addr >> 4;     // Head to the beginning of the completion region
+            cq_write_interface.completion_fifo_wr_toggle = not cq_write_interface.completion_fifo_wr_toggle;
+            notify_host_of_completion_queue_write_pointer<host_completion_queue_write_ptr_addr>();
+            noc_async_write_barrier(); // Barrier for now
+        } else if (reading_buffer) {
+            uint32_t buffer_transfer_start_addr = command_start_addr + (DeviceCommand::NUM_ENTRIES_IN_COMMAND_HEADER * sizeof(uint32_t));
+            volatile tt_l1_ptr uint32_t *buffer_transfer_command_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(buffer_transfer_start_addr);
+            db_cb_config_t* db_cb_config = get_local_db_cb_config(CQ_CONSUMER_CB_BASE, true);
+            const db_cb_config_t* eth_db_cb_config = get_remote_db_cb_config(eth_l1_mem::address_map::CQ_CONSUMER_CB_BASE, false);
+            uint32_t num_buffer_transfers = header->num_buffer_transfers;
+            bool is_sharded = (bool) (header->buffer_type == (uint32_t)DeviceCommand::BufferType::SHARDED);
+            uint32_t sharded_buffer_num_cores = header->sharded_buffer_num_cores;
+            uint32_t producer_consumer_transfer_num_pages = header->consumer_router_transfer_num_pages;
+            write_buffers(
+                db_cb_config,
+                eth_db_cb_config,
+                buffer_transfer_command_ptr,
+                num_buffer_transfers,
+                is_sharded,
+                sharded_buffer_num_cores,
+                producer_noc_encoding,
+                producer_consumer_transfer_num_pages);
+        }
+
+        uint32_t finish = header->finish;
+        if (finish) {
+            notify_host_complete<host_finish_addr>();
+        }
+
+        completion_queue_push_back<completion_queue_start_addr, host_completion_queue_write_ptr_addr>(completion_data_size);
+
+        // notify producer that it has completed a command
+        noc_semaphore_inc((uint64_t(producer_noc_encoding) << 32) | eth_get_semaphore(0), 1);
```

Comment:
I was thinking we should have some ENUMS mapped to the semaphores here. It isn't a problem for tensix cores, since every core that is doing dispatch has its own L1. But for erisc, if I want to add Semaphores for users, this will get confusing. I'm thinking we let users do `CreateSemaphore(0/1/2/3)` and for Dispatch, we can do `CreateSemaphore(FD_ETH_PRODUCER)`, `CreateSemaphore(FD_ETH_CONSUMER)`. No action here for now though.